scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/ClassDefinitionSetWithExtensionMethods.scala

```diff
@@ -0,0 +1,102 @@
+package pl.touk.nussknacker.engine.extension
+
+import cats.data.ValidatedNel
+import pl.touk.nussknacker.engine.ModelData
+import pl.touk.nussknacker.engine.api.Documentation
+import pl.touk.nussknacker.engine.api.generics.{GenericFunctionTypingError, MethodTypeInfo, Parameter}
+import pl.touk.nussknacker.engine.api.process.ClassExtractionSettings
+import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypingResult, Unknown}
+import pl.touk.nussknacker.engine.definition.clazz.{
+  ClassDefinition,
+  ClassDefinitionSet,
+  FunctionalMethodDefinition,
+  MethodDefinition
+}
+import pl.touk.nussknacker.engine.util.Implicits.RichScalaMap
+
+import java.lang.reflect.Method
+
+class ClassDefinitionSetWithExtensionMethods(set: ClassDefinitionSet, settings: ClassExtractionSettings) {
+  private val stringClass: Class[String] = classOf[String]
+  private val stringTyping: TypingResult = Typed.genericTypeClass(stringClass, Nil)
+  // We cannot have a class as the key because of `visibleMembersPredicate` e.g. Cast.castTo may be accessible for many classes.
+  // We only added statically defined methods because we needed a typingFunction which has access to some state.
+  // This functionality can be easily extended using dynamic methods discovery what you only need to do is:
+  // - Take all extension classes from class: ExtensionMethods, and then filter out those classes that are statically defined e.g. using annotation.
+  // - Filter out static and scala varargs methods (Method.isStatic and ClassDefinitionExtractor.javaVersionOfVarArgMethod)
+  // - Then invoke ClassDefinitionExtractor.extractMethod (you can easily build ClassDefinitionExtractor using settings).
+  // - After this we should filter out hidden members using ClassDefinitionExtractor.filterHiddenParameterAndReturnType.
+  // - And calculated definitions should be combined with static definitions.
+  private val extensionMethodsMap: Map[Method, Map[String, List[MethodDefinition]]] =
+    createStaticallyDefinedMethodsMap(set)
+
+  val value = new ClassDefinitionSet(
+    set.classDefinitionsMap.map { case (clazz, definition) =>
+      clazz -> enrichWithExtensionMethods(clazz, definition)
+    }.toMap // .toMap is needed by scala 2.12
+  )
+
+  private def enrichWithExtensionMethods(clazz: Class[_], classDefinition: ClassDefinition): ClassDefinition =
+    findMethodsForClass(clazz) match {
+      case ext if ext.isEmpty => classDefinition
+      case ext                => classDefinition.copy(methods = classDefinition.methods ++ ext)
+    }
+
+  private def findMethodsForClass(clazz: Class[_]): Map[String, List[MethodDefinition]] = {
+    val membersPredicate = settings.visibleMembersPredicate(clazz)
+    extensionMethodsMap
+      .filterKeysNow(membersPredicate.shouldBeVisible)
+      .flatMap(_._2)
+  }
+
+  private def createStaticallyDefinedMethodsMap(
```

Comment:
We won't need to this whole reflection magic with `castClass.getDeclaredMethod`, wrapping set with  `AllowedCastParametersClasses`, use internal `VisibleMembersPredicate` class of `ClassExtractionSettings` here, if we replace usage of `ClassExtractionSettings` with one simple rule described above. This rule could be placed near `Cast`/`CastTyping` to keep all casting logic in one place.

Reply:
Nice